Add support for bulk track/engage APIs

mbuhot opened the pull request:
- `Mixpanel.engage` function added to update user profile
- `Mixpanel.track` and `Mixpanel.engage` will both support a list of maps
- When given a list, these functions expect the maps to conform to the structure defined by
  the mixpanel HTTP api documentation (linked from README)

Commit:
Add support for bulk track/engage APIs

 - `Mixpanel.engage` function added to update user profile
 - `Mixpanel.track` and `Mixpanel.engage` will both support a list of maps
 - When given a list, these functions expect the maps to conform to the structure defined by
    the mixpanel HTTP api documentation (linked from README)

michihuber commented on lib/mixpanel.ex:
```diff
@@ -16,12 +16,27 @@ defmodule Mixpanel do
+    :ok
+  end
+
+  def engage(event) when is_map(event) do
```
Both `engage` fns have the same body, could you collapse them into a single fn, please?

michihuber commented on lib/mixpanel/client.ex:
```diff
@@ -18,19 +19,53 @@ defmodule Mixpanel.Client do
+    GenServer.cast(:mixpanel_client, {:track, events})
+  end
+
+  def engage(event) when is_map(event) do
```
These could also be the same fn.

michihuber commented on lib/mixpanel/client.ex:
```diff
@@ -18,19 +19,53 @@ defmodule Mixpanel.Client do
     properties = Dict.put(properties, :token, state.token)
-    {:ok, json} = JSX.encode([event: event, properties: properties])
-    body = String.to_char_list("data=#{ :base64.encode(json) }")
+    {:ok, json} = JSX.encode(event: event, properties: properties)
```
Does bulk posting a non-array work?
We should probably add some ex-vcr specs at this point.

michihuber commented on lib/mixpanel/client.ex:
```diff
@@ -18,19 +19,53 @@ defmodule Mixpanel.Client do
+
+  def handle_cast({:track, events}, state) when is_list(events) do
+    events = Enum.map(events, &put_in(&1, [:properties, :token], state.token))
+    {:ok, json} = JSX.encode(events)
```
could you dry up the encode/post/reply dance that's happening in all the `handle_cast`s, please?

michihuber reviewed (changes requested):
Hey Mike!

Thank you very much for the pull request!
I added some comments, mostly cosmetics, but I also think we should add some tests using ExVcr (which I should have done from the beginning, my bad).

If you want me to help out with the testing part, please let me know.

Thank you very much!

Commit:
DRY up implementation

 - Mixpanel.Client now only accepts lists of events
 - Mixpanel facade normalizes single events to lists of maps

Commit:
Add ExVCR tests

 - Note that these don't actually contain any asserts
 - And the ExVCR integration isn't working
 -- AFAICT the HTTPC adapter doesn't work with a custom :inets profile
 -- Even when using the default profile, the Mixpanel API is async, so need to wait in tests before asserting somehow

mbuhot commented:
Hey Michi, thanks for considering this PR.

I've tried to DRY up the implementation as per your feedback.
The `Mixpanel` module converts single events to lists and `Mixpanel.Client` only deals in lists of maps.

I didn't have much luck with ExVCR.
The only way I could get it to record was to change `GenServer.cast` to `GenServer.call` and to use `:httpc.request/4` instead of `:httpc.request/5` in `Mixpanel.Client`

I've added the mix dependency and left the test module in there, but it just exercises the code without actually asserting on anything right now.

mbuhot commented:
@michihuber Any suggestions on getting the VCRs on this one working?
